samza-test/src/test/java/org/apache/samza/test/processor/TestZkStreamProcessor.java

```diff
@@ -0,0 +1,558 @@
+
+    configs.put("job.systemstreampartition.grouper.factory",
+        "org.apache.samza.container.grouper.stream.GroupByPartitionFactory");
+    configs.put("task.name.grouper.factory", "org.apache.samza.container.grouper.task.GroupByContainerIdsFactory");
+
+    configs.put(JobCoordinatorConfig.JOB_COORDINATOR_FACTORY, "org.apache.samza.zk.ZkJobCoordinatorFactory");
+
+    return configs;
+  }
+
+  /**
+   * Produces the provided number of messages to the topic.
+   */
+  private void produceMessages(final int start, String topic, int numMessages) {
+    KafkaProducer producer = getKafkaProducer();
+    for (int i = start; i < numMessages + start; i++) {
+      try {
+        System.out.println("producing " + i);
+        producer.send(new ProducerRecord(topic, String.valueOf(i).getBytes())).get();
+      } catch (InterruptedException | ExecutionException e) {
+        e.printStackTrace();
+      }
+    }
+  }
+
+  /**
+   * Runs the provided stream processor by starting it, waiting on the provided latch with a timeout,
+   * and then stopping it.
+   */
+  private Thread runInThread(final StreamProcessor processor, CountDownLatch latch) {
+    Thread t = new Thread() {
+
+      @Override
+      public void run() {
+        processor.start();
+        try {
+          // just wait
+          synchronized (this) {
+            this.wait(100000);
+          }
+          System.out.println("notifed. Abandon the wait.");
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+        System.out.println("Stopping the processor");
+        processor.stop();
+      }
+    };
+    return t;
+  }
+
+  /**
+   * Consumes data from the topic until there are no new messages for a while
+   * and asserts that the number of consumed messages is as expected.
+   */
+  private void verifyNumMessages(String topic, final Map<Integer, Boolean> expectedValues, int expectedNumMessages) {
+    KafkaConsumer consumer = getKafkaConsumer();
+    consumer.subscribe(Collections.singletonList(topic));
+
+    Map<Integer, Boolean> map = new HashMap<>(expectedValues);
+    int count = 0;
+    int emptyPollCount = 0;
+
+    while (count < expectedNumMessages && emptyPollCount < 5) {
+      ConsumerRecords records = consumer.poll(5000);
+      if (!records.isEmpty()) {
+        Iterator<ConsumerRecord> iterator = records.iterator();
+        while (iterator.hasNext()) {
+          ConsumerRecord record = iterator.next();
+          //Assert.assertEquals(new String((byte[]) record.value()), String.valueOf(count));
+          String val = new String((byte[]) record.value());
+          System.out.println("Got value " + val);
+          map.put(Integer.valueOf(val), true);
+          count++;
+        }
+      } else {
+        emptyPollCount++;
+        System.out.println("empty polls " + emptyPollCount);
+      }
+    }
+    // filter out numbers we did not get
+    long numFalse = map.values().stream().filter(v -> !v).count();
```

Comment:
as I said the list of expected values may change based on the test. So decision what are the values, should be done in the test itself. I can create an auxiliary method to make it look nicer.